File: FSharpWithAkka/Account/Domain/Account.fs
[<RequireQualifiedAccess>]
module Account

open System

open BankTypes
open Bank.Account.Domain
open Bank.Transfer.Domain
open Lib.Types
open Lib.Time

let streamName (id: Guid) = "accounts_" + string id

let recipientLookupKey (recipient: TransferRecipient) =
   match recipient.AccountEnvironment with
   | RecipientAccountEnvironment.Internal -> recipient.Identification
   | RecipientAccountEnvironment.Domestic ->
      $"{recipient.RoutingNumber.Value}_{recipient.Identification}"

let DailyDebitAccrued state (evt: BankEvent<DebitedAccount>) : decimal =
   // When accumulating events into AccountState aggregate...
   // -> Ignore debits older than a day
   if not <| IsToday evt.Timestamp then
      0m
   elif Option.isNone state.LastDebitDate then
      evt.Data.DebitedAmount
   // When applying a new event to the cached AccountState & the
   // last debit event did not occur today...
   // -> Ignore the cached DailyDebitAccrued
   elif not <| IsToday state.LastDebitDate.Value then
      evt.Data.DebitedAmount
   else
      state.DailyDebitAccrued + evt.Data.DebitedAmount

let create (e: BankEvent<CreatedAccount>) = {
   EntityId = e.EntityId
   FirstName = e.Data.FirstName
   LastName = e.Data.LastName
   Currency = e.Data.Currency
   Balance = e.Data.Balance
   Status = AccountStatus.Active
   AllowedOverdraft = 0m
   DailyDebitLimit = -1m
   DailyDebitAccrued = 0m
   LastDebitDate = None
   TransferRecipients = Map.empty
}

let applyEvent (state: AccountState) (evt: AccountEvent) =
   match evt with
   | DepositedCash e -> {
      state with
         Balance = state.Balance + e.Data.DepositedAmount
     }
   | DebitedAccount e -> {
      state with
         Balance = state.Balance - e.Data.DebitedAmount
         DailyDebitAccrued = DailyDebitAccrued state e
         LastDebitDate = Some e.Timestamp
     }
   | MaintenanceFeeDebited e -> {
      state with
         Balance = state.Balance - e.Data.DebitedAmount
     }
   | DailyDebitLimitUpdated e -> {
      state with
         DailyDebitLimit = e.Data.DebitLimit
     }
   | LockedCard _ -> {
      state with
         Status = AccountStatus.ActiveWithLockedCard
     }
   | UnlockedCard _ -> {
      state with
         Status = AccountStatus.Active
     }
   | TransferPending e -> {
      state with
         Balance = state.Balance - e.Data.DebitedAmount
     }
   | TransferApproved _ -> state
   | TransferRejected e -> {
      state with
         Balance = state.Balance + e.Data.DebitedAmount
     }
   | InternalTransferRecipient e ->
      let recipient =
         RegisterTransferRecipientEvent.eventToRecipient (
            e |> RegisteredInternalTransferRecipient
         )

      let key = recipientLookupKey recipient

      {
         state with
            TransferRecipients = state.TransferRecipients.Add(key, recipient)
      }
   | DomesticTransferRecipient e ->
      let recipient =
         RegisterTransferRecipientEvent.eventToRecipient (
            e |> RegisteredDomesticTransferRecipient
         )

      let key = recipientLookupKey recipient

      {
         state with
            TransferRecipients = state.TransferRecipients.Add(key, recipient)
      }
   | InternationalTransferRecipient e ->
      let recipient =
         RegisterTransferRecipientEvent.eventToRecipient (
            e |> RegisteredInternationalTransferRecipient
         )

      let key = recipientLookupKey recipient

      {
         state with
            TransferRecipients = state.TransferRecipients.Add(key, recipient)
      }
   | _ -> state

module private StateTransition =
   let deposit (state: AccountState) (cmd: DepositCashCommand) =
      if state.Status = AccountStatus.Closed then
         Error "AccountNotActive"
      elif cmd.Amount <= 0m then
         Error "InvalidDepositAmount"
      else
         let evt = DepositedCashEvent.create cmd |> DepositedCash
         Ok(evt, applyEvent state evt)

   let limitDailyDebits (state: AccountState) (cmd: LimitDailyDebitsCommand) =
      let evt = DailyDebitLimitUpdatedEvent.create cmd |> DailyDebitLimitUpdated

      Ok(evt, applyEvent state evt)

   let lockCard (state: AccountState) (cmd: LockCardCommand) =
      if state.Status <> AccountStatus.Active then
         Error "AccountNotActive"
      else
         let evt = LockedCardEvent.create cmd |> LockedCard
         Ok(evt, applyEvent state evt)

   let unlockCard (state: AccountState) (cmd: UnlockCardCommand) =
      if state.Status <> AccountStatus.ActiveWithLockedCard then
         Error $"Account card already unlocked {state.Status}"
      else
         let evt = UnlockedCardEvent.create cmd |> UnlockedCard
         Ok(evt, applyEvent state evt)

   let debit (state: AccountState) (cmd: DebitCommand) =
      if state.Status = AccountStatus.Closed then
         Error "AccountNotActive"
      elif state.Status = AccountStatus.ActiveWithLockedCard then
         Error "AccountCardLocked"
      elif state.Balance - cmd.Amount < state.AllowedOverdraft then
         Error "InsufficientBalance"
      elif
         state.DailyDebitLimit <> -1m
         && IsToday cmd.Timestamp
         && state.DailyDebitAccrued + cmd.Amount > state.DailyDebitLimit
      then
         Error $"ExceededDailyDebit {state.DailyDebitLimit}"
      else
         let evt = DebitedAccountEvent.create cmd |> DebitedAccount
         Ok(evt, applyEvent state evt)

   let maintenanceFee (state: AccountState) (cmd: MaintenanceFeeCommand) =
      if state.Status = AccountStatus.Closed then
         Error "AccountNotActive"
      elif state.Balance - cmd.Amount < state.AllowedOverdraft then
         Error "InsufficientBalance"
      else
         let evt = MaintenanceFeeEvent.create cmd |> MaintenanceFeeDebited
         Ok(evt, applyEvent state evt)

   let transfer (state: AccountState) (cmd: TransferCommand) =
      if state.Status = AccountStatus.Closed then
         Error "AccountNotActive"
      elif state.Balance - cmd.Amount < state.AllowedOverdraft then
         Error "InsufficientBalance"
      elif
         not
         <| state.TransferRecipients.ContainsKey(
            recipientLookupKey cmd.Recipient
         )
      then
         Error "TransferErr.RecipientRegistrationRequired(cmd)"
      else
         let evt = TransferEvent.create cmd |> TransferPending
         Ok(evt, applyEvent state evt)

   let approveTransfer (state: AccountState) (cmd: ApproveTransferCommand) =
      let evt = TransferEvent.approve cmd |> TransferApproved

      Ok(evt, applyEvent state evt)

   let rejectTransfer (state: AccountState) (cmd: RejectTransferCommand) =
      let evt = TransferEvent.reject cmd |> TransferRejected

      Ok(evt, applyEvent state evt)

   let registerTransferRecipient
      (state: AccountState)
      (cmd: RegisterTransferRecipientCommand)
      =
      if state.TransferRecipients.ContainsKey cmd.Recipient.Identification then
         Error "TransferErr.RecipientAlreadyRegistered(cmd)"
      else
         let evt =
            match cmd.Recipient.AccountEnvironment with
            | RecipientAccountEnvironment.Internal ->
               RegisterInternalTransferRecipientEvent.create cmd
               |> InternalTransferRecipient
            | RecipientAccountEnvironment.Domestic ->
               RegisterDomesticTransferRecipientEvent.create cmd
               |> DomesticTransferRecipient
            | RecipientAccountEnvironment.International ->
               RegisterInternationalTransferRecipientEvent.create cmd
               |> InternationalTransferRecipient

         Ok(evt, applyEvent state evt)

let stateTransition (state: AccountState) (command: Command) =
   match box command with
   | :? DepositCashCommand as cmd -> StateTransition.deposit state cmd
   | :? DebitCommand as cmd -> StateTransition.debit state cmd
   | :? MaintenanceFeeCommand as cmd -> StateTransition.maintenanceFee state cmd
   | :? LimitDailyDebitsCommand as cmd ->
      StateTransition.limitDailyDebits state cmd
   | :? LockCardCommand as cmd -> StateTransition.lockCard state cmd
   | :? UnlockCardCommand as cmd -> StateTransition.unlockCard state cmd
   | :? TransferCommand as cmd -> StateTransition.transfer state cmd
   | :? ApproveTransferCommand as cmd ->
      StateTransition.approveTransfer state cmd
   | :? RejectTransferCommand as cmd -> StateTransition.rejectTransfer state cmd
   | :? RegisterTransferRecipientCommand as cmd ->
      StateTransition.registerTransferRecipient state cmd

let initialAccountStateFromEventHistory events =
   let (CreatedAccount createdEvt) = List.head events
   create createdEvt

let foldEventsIntoAccount events =
   List.fold
      applyEvent
      (initialAccountStateFromEventHistory events)
      (List.tail events)

File: Archive/CSharpWithLanguageExt/Account/Domain/Account.cs
using LanguageExt;

using Time = Lib.Time;
using Lib.Types;
using Lib.BankTypes;
using Bank.Transfer.Domain;
using MF = Bank.MaintenanceFee.Domain.MaintenanceFee;

namespace Bank.Account.Domain;

using StateTransitionResult = Validation<Err, (Event Event, AccountState NewState)>;

public static class Account {
   public static string StreamName(Guid id) => $"accounts_{id}";

   public static Map<string, Type> EventTypeMapping = (
      (nameof(CreatedAccount), typeof(CreatedAccount)),
      (nameof(TransferPending), typeof(TransferPending)),
      (nameof(TransferApproved), typeof(TransferApproved)),
      (nameof(TransferRejected), typeof(TransferRejected)),
      (nameof(DebitedAccount), typeof(DebitedAccount)),
      (nameof(MaintenanceFeeDebited), typeof(MaintenanceFeeDebited)),
      (nameof(MaintenanceFeeSkipped), typeof(MaintenanceFeeSkipped)),
      (nameof(DailyDebitLimitUpdated), typeof(DailyDebitLimitUpdated)),
      (nameof(DepositedCash), typeof(DepositedCash)),
      (nameof(LockedCard), typeof(LockedCard)),
      (nameof(UnlockedCard), typeof(UnlockedCard)),
      (nameof(RegisteredInternalTransferRecipient), typeof(RegisteredInternalTransferRecipient)),
      (nameof(RegisteredDomesticTransferRecipient), typeof(RegisteredDomesticTransferRecipient)),
      (nameof(RegisteredInternationalTransferRecipient), typeof(RegisteredInternationalTransferRecipient))
   );

   public static AccountState Apply(this AccountState acc, object evt) =>
      evt switch {
         DepositedCash e =>
            MF.FromDeposit(
               acc with { Balance = acc.Balance + e.DepositedAmount },
               e.DepositedAmount),

         TransferPending e =>
            MF.FromDebit(acc with { Balance = acc.Balance - e.DebitedAmount }),

         TransferApproved _ => acc,

         TransferRejected e =>
            MF.FromDebitReversal(acc with { Balance = acc.Balance + e.DebitedAmount }),

         MaintenanceFeeDebited e =>
            MF.Reset(acc with { Balance = acc.Balance - e.DebitedAmount }),

         MaintenanceFeeSkipped _ => MF.Reset(acc),

         DebitedAccount e =>
            MF.FromDebit(acc with {
               Balance = acc.Balance - e.DebitedAmount,
               DailyDebitAccrued = DailyDebitAccrued(acc, e),
               LastDebitDate = e.Timestamp
            }),

         DailyDebitLimitUpdated e =>
            acc with { DailyDebitLimit = e.DebitLimit },

         LockedCard _ =>
            acc with { Status = AccountStatus.ActiveWithLockedCard },

         UnlockedCard _ =>
            acc with { Status = AccountStatus.Active },

         RegisteredInternalTransferRecipient e => acc with {
            TransferRecipients = acc.TransferRecipients.AddOrUpdate(
               e.AccountNumber,
               e.AsTransferRecipient()
            )
         },

         RegisteredDomesticTransferRecipient e => acc with {
            TransferRecipients = acc.TransferRecipients.AddOrUpdate(
               $"{e.RoutingNumber}_{e.AccountNumber}",
               e.AsTransferRecipient()
            )
         },

         RegisteredInternationalTransferRecipient e => acc with {
            TransferRecipients = acc.TransferRecipients.AddOrUpdate(
               e.Identification,
               e.AsTransferRecipient()
            )
         },

         _ => acc
      };

   public static Decimal DailyDebitAccrued(AccountState state, DebitedAccount evt) {
      // When accumulating events into AccountState aggregate...
      // -> Ignore debits older than a day
      if (!Time.IsToday(evt.Timestamp)) return 0;

      // When applying a new event to the cached AccountState & the
      // last debit event did not occur today...
      // -> Ignore the cached DailyDebitAccrued
      if (!Time.IsToday(state.LastDebitDate)) return evt.DebitedAmount;

      return state.DailyDebitAccrued + evt.DebitedAmount;
   }

   public static StateTransitionResult StateTransition(
      this AccountState state,
      Command command
   ) =>
      command switch {
         TransferCmd cmd => state.Transfer(cmd),
         ApproveTransferCmd cmd => state.ApproveTransfer(cmd),
         RejectTransferCmd cmd => state.RejectTransfer(cmd),
         DebitCmd cmd => state.Debit(cmd),
         MaintenanceFeeCmd cmd => state.MaintenanceFee(cmd),
         SkipMaintenanceFeeCmd cmd => state.SkipMaintenanceFee(cmd),
         LimitDailyDebitsCmd cmd => state.LimitDailyDebits(cmd),
         DepositCashCmd cmd => state.Deposit(cmd),
         LockCardCmd cmd => state.LockCard(cmd),
         UnlockCardCmd cmd => state.UnlockCard(cmd),
         RegisterTransferRecipientCmd cmd => state.RegisterTransferRecipient(cmd),
         _ => new Err("State Transition not implemented for command.")
      };

   public static StateTransitionResult Transfer(
      this AccountState state,
      TransferCmd cmd
   ) {
      if (state.Status == AccountStatus.Closed)
         return Errors.AccountNotActive;

      if (state.Balance - cmd.Amount < state.AllowedOverdraft)
         return Errors.InsufficientBalance;

      if (state.TransferRecipients.Find(cmd.Recipient.Identification).IsNone)
         return TransferErr.RecipientRegistrationRequired(cmd);

      var evt = cmd.ToEvent();
      return (evt, state.Apply(evt));
   }

   public static StateTransitionResult ApproveTransfer(
      this AccountState state,
      ApproveTransferCmd cmd
   ) {
      var evt = cmd.ToEvent();
      return (evt, state.Apply(evt));
   }

   public static StateTransitionResult RejectTransfer(
      this AccountState state,
      RejectTransferCmd cmd
   ) {
      var evt = cmd.ToEvent();
      return (evt, state.Apply(evt));
   }

   public static StateTransitionResult MaintenanceFee(
      this AccountState state,
      MaintenanceFeeCmd cmd
   ) {
      if (state.Status == AccountStatus.Closed)
         return Errors.AccountNotActive;
      if (state.Balance - cmd.Amount < state.AllowedOverdraft)
         return Errors.InsufficientBalance;

      var evt = cmd.ToEvent();
      return (evt, state.Apply(evt));
   }

   public static StateTransitionResult SkipMaintenanceFee(
      this AccountState state,
      SkipMaintenanceFeeCmd cmd
   ) {
      var evt = cmd.ToEvent();
      return (evt, state.Apply(evt));
   }

   public static StateTransitionResult Debit(
      this AccountState state,
      DebitCmd cmd
   ) {
      if (state.Status == AccountStatus.Closed)
         return Errors.AccountNotActive;

      if (state.Status == AccountStatus.ActiveWithLockedCard)
         return Errors.AccountCardLocked;

      if (state.Balance - cmd.Amount < state.AllowedOverdraft)
         return Errors.InsufficientBalance;

      if (state.DailyDebitLimit != -1 &&
         Time.IsToday(cmd.Timestamp) &&
         state.DailyDebitAccrued + cmd.Amount > state.DailyDebitLimit
      ) {
         return Errors.ExceededDailyDebitLimit(state.DailyDebitLimit);
      }

      var evt = cmd.ToEvent();
      return (evt, state.Apply(evt));
   }

   public static StateTransitionResult LimitDailyDebits(
      this AccountState state,
      LimitDailyDebitsCmd cmd
   ) {
      var evt = cmd.ToEvent();
      return (evt, state.Apply(evt));
   }

   public static StateTransitionResult Deposit(
      this AccountState state,
      DepositCashCmd cmd
   ) {
      if (state.Status == AccountStatus.Closed)
         return Errors.AccountNotActive;

      if (cmd.Amount <= 0)
         return Errors.InvalidDepositAmount;

      var evt = cmd.ToEvent();
      return (evt, state.Apply(evt));
   }

   public static StateTransitionResult LockCard(
      this AccountState state,
      LockCardCmd cmd
   ) {
      if (state.Status != AccountStatus.Active)
         return Errors.AccountNotActive;

      var evt = cmd.ToEvent();
      return (evt, state.Apply(evt));
   }

   public static StateTransitionResult UnlockCard(
      this AccountState state,
      UnlockCardCmd cmd
   ) {
      if (state.Status != AccountStatus.ActiveWithLockedCard)
         return new Err($"Account card already unlocked: {state.Status}");

      var evt = cmd.ToEvent();
      return (evt, state.Apply(evt));
   }

   public static StateTransitionResult RegisterTransferRecipient(
      this AccountState state,
      RegisterTransferRecipientCmd cmd
   ) {
      if (state.TransferRecipients.Find(cmd.Recipient.Identification).IsSome)
         return TransferErr.RecipientAlreadyRegistered(cmd);

      var evt = cmd.ToEvent().Unwrap();
      return (evt, state.Apply(evt));
   }

   public static AccountState Create(CreatedAccount evt) =>
      MF.Reset(new AccountState(
         FirstName: evt.FirstName,
         LastName: evt.LastName,
         EntityId: evt.EntityId,
         Currency: evt.Currency,
         Balance: evt.Balance,
         MaintenanceFeeCriteria: new()
      ));
}
